weblate/api/serializers.py

```diff
@@ -156,6 +167,24 @@ class Meta:
         )
         extra_kwargs = {"url": {"view_name": "api:role-detail", "lookup_field": "id"}}
 
+    def create(self, validated_data):
+        permissions_validated = validated_data.pop("permissions", [])
+        role = Role.objects.create(**validated_data)
+        for codename in permissions_validated:
+            permission = Permission.objects.get(codename=codename)
+            role.permissions.add(permission)
+        return role
+
+    def update(self, instance, validated_data):
+        permissions_validated = validated_data.pop("permissions", [])
+        instance.name = validated_data.get("name", instance.name)
+        instance.save()
+        instance.permissions.clear()
+        for codename in permissions_validated:
+            permission = Permission.objects.get(codename=codename)
+            instance.permissions.add(permission)
```

Comment:
This block could use similar simplification as in create, but there is additional concern - this will only *add* permisssions, but there is no way to remove them. Probably `set()` should be used here instead of `add()`, but interaction with partial updates has to be defined and tested. In case PATCH request is missing permissions, it should remain untouched.

Reply:
@nijel I did the `instance.permissions.clear()` because of that. So as to remove them. Right now, yes, PATCH will need to have all permissions, missing permissions will get removed. So need to fix that. I was thinking of adding a separate definition for partial_update in the viewset for PATCH.

Reply:
Ah, I didn't notice the `clear()` invocation. Anyway `set()` is better approach here as it avoids removing and adding back not changed permissions.

As for PATCH, you can figure out whether it is being done by looking at `self.partial` here.